python/grass/script/setup.py

```diff
@@ -268,7 +268,7 @@ def setup_runtime_env(gisbase=None, *, env=None):
     env["PYTHONPATH"] = path
 
 
-def init(path, location=None, mapset=None, grass_path=None):
+def init(path, location=None, mapset=None, *, grass_path=None, env=None):
```

Comment:
This should have triggered a warning with a Python linter. Usually positional arguments shouldn't be after named (keyword) arguments. Was this thinked out and must be this for keeping the existing signature ? If so, why is there another existing keyword argument after?

Reply:
grass_path and env are now [PEP 3102 – Keyword-Only Arguments](https://peps.python.org/pep-3102/).

The thing is, though, this is actually slightly changing the API introduced in 8.0.0. The following was possible before:

```python
gs.setup.init("~/data", "nc_spm", "new_highway", "/opt/grass/bin/grass")
```

The above now must be rewritten as:

```python
gs.setup.init("~/data", "nc_spm", "new_highway", grass_path="/opt/grass/bin/grass")
```

However, you had to use _grass_path_ explicitly even before when passing full path to mapset:

```python
gs.setup.init("~/data/nc_spm/new_highway", grass_path="/opt/grass/bin/grass")
```

And that how the "basic usage" in the documentation for 8.0.0 shows gs.setup.init is to be used:

```python
import grass.script as gs
session = gs.setup.init(
    "~/grassdata/nc_spm_08/user1",
    grass_path="/usr/lib/grass",
)
```

Importantly, on most systems, the heuristic in the function determines the path automatically, so _grass_path_ does not have to be supplied at all. None of our notebook examples use _grass_path_.

Strictly following semantic versioning, the asterisk should be before _env_, but putting it before _grass_path_ codifies the intended usage. Why I did not include it into the original API, I don't know. Maybe I didn't know about it...Maybe we were still aiming at Python 2 compatibility...Maybe some other controversy...

Reply:
How do you feel about putting the asterisk before _grass_path_ when strict API compatibility would dictate to put it only after _grass_path_ and before _env_? I'm for breaking the strict API compatibility and going with what the documentation suggests which is using _grass_path_ a keyword argument. I also think this is low risk because this is a new API.

Reply:
Was it realeased in 8.3?

Reply:
Without actually checking this, I think the new init parameters were part of 8.0 (because before, they were very different).

Reply:
I still can't decide myself on what to do with this.
Was it ever possible to have other positional arguments that need to have api compatibility?

Reply:
I don't understand the question. Can you please give me an example of what you mean?

Reply:
I was trying to see if it was really necessary to try and keep API compatibility. If before there wasn't any additional positional arguments (`*`), nor extra keyword-only arguments `**kwargs`, how would've this function be could been called in a way that having the `*` placed before or after matters? If it was impossible to call in a different way, we don't need to overthink, and just choose the correct option for the future.

Reply:
It is not about extra arguments, it is about any arguments, in this case, mainly about keyword arguments:

Worked in 8.0-8.3:

```python
gs.setup.init("~/data/nc_spm/new_highway")
gs.setup.init("~/data/nc_spm/new_highway", grass_path="/opt/grass/bin/grass")
gs.setup.init("~/data", "nc_spm", "new_highway")
gs.setup.init("~/data", "nc_spm", "new_highway", "/opt/grass/bin/grass")
```

With this PR:

```python
gs.setup.init("~/data/nc_spm/new_highway")
gs.setup.init("~/data/nc_spm/new_highway", grass_path="/opt/grass/bin/grass")
gs.setup.init("~/data", "nc_spm", "new_highway")
gs.setup.init("~/data/nc_spm/new_highway", env=env)
gs.setup.init("~/data/nc_spm/new_highway", grass_path="/opt/grass/bin/grass", env=env)
gs.setup.init("~/data", "nc_spm", "new_highway", env=env)
```

Never possible:


```python
gs.setup.init("~/data/nc_spm/new_highway", "/opt/grass/bin/grass")
```

With this PR, code without an explicit _grass_path_ working in 8.0-8.3 needs to be rewritten:

```diff
- gs.setup.init("~/data", "nc_spm", "new_highway", "/opt/grass/bin/grass")
+ gs.setup.init("~/data", "nc_spm", "new_highway", grass_path="/opt/grass/bin/grass")
```

Back to my earlier comment, the usage which is no longer available with this PR was never promoted. Instead, the documentation for _grass.script.setup.init_ shows:

```python
import grass.script as gs
session = gs.setup.init(
    "~/grassdata/nc_spm_08/user1",
    grass_path="/usr/lib/grass",
)
```

There is a heuristic in the function which determines the path automatically, so the idea is that in most cases, _grass_path_ does not have to be supplied at all. (I don't see any usages of it [on GitHub](https://github.com/search?q=%22grass_path%3D%22&type=code).) Having _grass_path_ passed as a keyword argument (with explicit name), rather than a positional one, is the intended usage (based on the example, number and type of parameters and my judgement of what I intended back in 8.0 times).

One can say that this PR limits the usage to intended usage by fixing bug in the function specification.

Reply:
I think it's ok then, especially with the other PR